lib/backend/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,325 @@
+// returns an error if the key does not exist or the data is empty.
+func (b *Backend) readSecretData(ctx context.Context, key []byte) (*backend.Item, error) {
+	secret, err := b.getSecret(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	data, ok := secret.Data[backendKeyToSecret(key)]
+	if !ok || len(data) == 0 {
+		return nil, trace.NotFound("key [%s] not found in secret %s", string(key), b.secretName)
+	}
+
+	return &backend.Item{
+		Key:   key,
+		Value: data,
+	}, nil
+}
+
+func (b *Backend) updateSecretContent(ctx context.Context, items ...backend.Item) (*backend.Lease, error) {
+	// FIXME(tigrato):
+	// for now, the agent is the owner of the secret so it's safe to replace changes
+
+	secret, err := b.getSecret(ctx)
+	if trace.IsNotFound(err) {
+		secret = b.genSecretObject()
+	} else if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if secret.Data == nil {
+		secret.Data = map[string][]byte{}
+	}
+
+	updateDataMap(secret.Data, items...)
+
+	if err := b.upsertSecret(ctx, secret); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &backend.Lease{}, nil
+}
+
+func (b *Backend) upsertSecret(ctx context.Context, secret *corev1.Secret) error {
+	secretApply := applyconfigv1.Secret(b.secretName, b.namespace).
+		WithData(secret.Data).
+		WithLabels(secret.GetLabels()).
+		WithAnnotations(secret.GetAnnotations())
+
+	// apply resource lock if it's not a creation
+	if len(secret.ResourceVersion) > 0 {
+		secretApply = secretApply.WithResourceVersion(secret.ResourceVersion)
+	}
+
+	_, err := b.k8sClientSet.
+		CoreV1().
+		Secrets(b.namespace).
+		Apply(ctx, secretApply, metav1.ApplyOptions{FieldManager: b.replicaName})
+
+	return trace.Wrap(err)
+}
+
+func (b *Backend) genSecretObject() *corev1.Secret {
+	return &corev1.Secret{
+		Type: corev1.SecretTypeOpaque,
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        b.secretName,
+			Namespace:   b.namespace,
+			Annotations: generateSecretAnnotations(b.namespace, b.releaseName),
+		},
+		Data: map[string][]byte{},
+	}
+
+}
+
+func generateSecretAnnotations(namespace, releaseNameEnv string) map[string]string {
+	const (
+		helmReleaseNameAnnotation     = "meta.helm.sh/release-name"
+		helmReleaseNamesaceAnnotation = "meta.helm.sh/release-namespace"
+		helmResourcePolicy            = "helm.sh/resource-policy"
+	)
+
+	if len(releaseNameEnv) > 0 {
+		return map[string]string{
+			helmReleaseNameAnnotation:     releaseNameEnv,
+			helmReleaseNamesaceAnnotation: namespace,
+			helmResourcePolicy:            "keep",
+		}
+	}
+
+	return map[string]string{}
+}
+
+// backendKeyToSecret replaces the "/" with "."
+// "/" chars are not allowed in Kubernetes Secret keys.
+func backendKeyToSecret(k []byte) string {
+	return strings.ReplaceAll(string(k), "/", ".")
+}
+
+func updateDataMap(data map[string][]byte, items ...backend.Item) {
+	for _, item := range items {
+		data[backendKeyToSecret(item.Key)] = item.Value
+	}
+}
+
+func validateConfig(c Config) error {
```

Comment:
Nit: Usually the pattern we use is to define a `Check()` method on the config struct that validates it. I would do the same here for consistency.